### PR TITLE
Fix issues with GitOps Service when running against KCP Virtual Workspace environment

### PR DESCRIPTION
**Pull request by chetan-rns**
#### Description:
- Use separate clients for accessing resources in the service provider and user workspaces.
- Update context to include cluster name parameter.

https://issues.redhat.com/browse/GITOPSRVCE-208

### CODE REVIEW

File: backend/eventloop/application_event_loop/application_event_runner.go [hunk 1]
```diff
@@ -97,6 +98,10 @@ func applicationEventLoopRunner(inputChannel chan *eventlooptypes.EventLoopEvent
 
 		ctx, cancel := context.WithCancel(outerContext)
 
+		if newEvent.Request.ClusterName != "" && !sharedutil.IsKCPVirtualWorkspaceDisabled() {
```
Comment (jgwest) [inline backend/eventloop/application_event_loop/application_event_runner.go:101]: Looks like we're repeating this snippet of code a few times, is it worth it to create a function for it in shared util, something like:
```go
	func addKCPClusterToContext(ctx context.Context, string clusterName) context.Context {
		if clusterName != "" && !sharedutil.IsKCPVirtualWorkspaceDisabled() {
			ctx = logicalcluster.WithCluster(ctx, logicalcluster.New(newEvent.Request.clusterName))
			return ctx
		} else {
			return ctx
		}
	}
```
Then calling it like
```go
	ctx = addKCPClusterToContext(ctx, newEvent.Request.ClusterName)
```

**Review by jgwest — CHANGES_REQUESTED** [review]
Looks great @chetan-rns, a couple quick requested changes, and a question.

I'm seeing an error in Argo CD `Application` resource status field when testing the ability to deploy using a ManagedEnvironment targeting the user's KCP workspace:  
```
Error from server (Forbidden): clusterworkspaces.tenancy.kcp.dev is forbidden: User "rh-sso:jgwest-kcp" cannot list resource "clusterworkspaces" in API group "tenancy.kcp.dev" at the cluster scope
```

It looks like we need to exclude `tenancy.kcp.dev` from the list of resources that Argo CD tries to watch, [like they had to do in the infra-deployments repo](https://github.com/redhat-appstudio/infra-deployments/blob/2ec6897a7b1d640c2a42af4d6601e8cd5148a5ad/hack/bootstrap.sh#L107).

Did you hit this, and/or were you able to get it to work? (If not, we can open a new story for this).


### Steps I used to test, to get this error:
<details>

```bash
git clone git@github.com:chetan-rns/managed-gitops
cd managed-gitops
git checkout verify-virtual-ws

# CPS stable
export CPS_KUBECONFIG=/home/jgw/.kube/kcp-config
export WORKLOAD_KUBECONFIG=/home/jgw/.kube/config

kcp/virtual-workspace.sh


# In another window, after above:

# switch to a separate kcp config for user
export CPS_KUBECONFIG=$HOME/.kube/kcp-config
cp $CPS_KUBECONFIG $HOME/.kube/kcp-config-user
export KUBECONFIG=$HOME/.kube/kcp-config-user

kubectl ws 

kubectl ws user-(user id)

kubectl create namespace jane
kubens jane

kubectl create secret generic my-managed-environment-secret \
       --from-file=kubeconfig=$HOME/.kube/kcp-config-user \
       --type="managed-gitops.redhat.com/managed-environment" \
       -n jane

cd examples/m6-demo

# update my-managed-environment.yaml to point to API URL from $HOME/.kube/kcp-config-user
kubectl apply -f my-managed-environment.yaml

kubectl apply -f gitops-deployment-managed-environment.yaml
```
</details>


### Full Error from Argo CD Application resource:
<details>

```yaml
apiVersion: v1
items:
- apiVersion: argoproj.io/v1alpha1
  kind: Application
  metadata:
    annotations:
      kcp.dev/cluster: root:users:oq:hx:rh-sso-jgwest-kcp:service-10024
    creationTimestamp: "2022-10-16T20:46:06Z"
    generation: 3
    labels:
      databaseID: 2f58de2d-7a32-416c-9eef-97bc54f3abba
      state.workload.kcp.dev/7b5Fc0uSqAciGBgLasbLOAnC825aDbSBA7F8E2: Sync
    name: gitopsdepl-e8cd4392-0bf5-42b5-a9d1-edf6edd6d311
    namespace: gitops-service-argocd
    resourceVersion: "215132343"
    uid: 549e229c-73f1-4177-a6d8-947b7332d615
  spec:
    destination:
      name: managed-env-2fce64a2-5025-4729-95bb-5cee3494c1d3
      namespace: jane
    project: default
    source:
      path: environments/overlays/dev
      repoURL: https://github.com/redhat-appstudio/gitops-repository-template
    syncPolicy:
      automated:
        allowEmpty: true
        prune: true
        selfHeal: true
      syncOptions:
      - PrunePropagationPolicy=background
  status:
    conditions:
    - lastTransitionTime: "2022-10-16T20:46:06Z"
      message: 'failed to sync cluster https://kcp-stable.apps.kcp-internal-1.52p3.p1.openshiftapps.com:443/clusters/root:users:oq:hx:rh-sso-jgwest-kcp:user-29347:
        failed to load initial state of resource ClusterWorkspace.tenancy.kcp.dev:
        clusterworkspaces.tenancy.kcp.dev is forbidden: User "system:serviceaccount:kube-system:argocd-manager-ae2b7716-9c49-442d-b460-58f5d758ec16"
        cannot list resource "clusterworkspaces" in API group "tenancy.kcp.dev" at
        the cluster scope'
      type: ComparisonError
    - lastTransitionTime: "2022-10-16T20:46:06Z"
      message: 'failed to sync cluster https://kcp-stable.apps.kcp-internal-1.52p3.p1.openshiftapps.com:443/clusters/root:users:oq:hx:rh-sso-jgwest-kcp:user-29347:
        failed to load initial state of resource ClusterWorkspace.tenancy.kcp.dev:
        clusterworkspaces.tenancy.kcp.dev is forbidden: User "system:serviceaccount:kube-system:argocd-manager-ae2b7716-9c49-442d-b460-58f5d758ec16"
        cannot list resource "clusterworkspaces" in API group "tenancy.kcp.dev" at
        the cluster scope'
      type: ComparisonError
    - lastTransitionTime: "2022-10-16T20:46:06Z"
      message: 'failed to sync cluster https://kcp-stable.apps.kcp-internal-1.52p3.p1.openshiftapps.com:443/clusters/root:users:oq:hx:rh-sso-jgwest-kcp:user-29347:
        failed to load initial state of resource ClusterWorkspace.tenancy.kcp.dev:
        clusterworkspaces.tenancy.kcp.dev is forbidden: User "system:serviceaccount:kube-system:argocd-manager-ae2b7716-9c49-442d-b460-58f5d758ec16"
        cannot list resource "clusterworkspaces" in API group "tenancy.kcp.dev" at
        the cluster scope'
      type: UnknownError
    health:
      status: Healthy
    reconciledAt: "2022-10-16T20:46:06Z"
    summary: {}
    sync:
      comparedTo:
        destination:
          name: managed-env-2fce64a2-5025-4729-95bb-5cee3494c1d3
          namespace: jane
        source:
          path: environments/overlays/dev
          repoURL: https://github.com/redhat-appstudio/gitops-repository-template
      status: Unknown
kind: List
metadata:
  resourceVersion: ""
```
</details>

**Comment by chetan-rns** [comment]
> Error from server (Forbidden): clusterworkspaces.tenancy.kcp.dev is forbidden: User "rh-sso:jgwest-kcp" cannot list resource "clusterworkspaces" in API group "tenancy.kcp.dev" at the cluster scope

@jgwest  I did hit this error while testing the PR. I found that the Argo CD cluster was configured with `ALL_NAMESPACES`  even though Argo CD was installed in namespace mode. This could be the reason behind the error: "cannot list resource at cluster scope"

So, I manually included the target namespace e.g jane in the cluster secret and the error was resolved.

File: backend/eventloop/application_event_loop/application_event_runner.go (hunk 1 above)
Comment (chetan-rns) [reply backend/eventloop/application_event_loop/application_event_runner.go]: Updated

File: backend/eventloop/application_event_loop/application_event_runner_deployments.go [hunk 2]
```diff
@@ -297,15 +297,15 @@ func (a applicationEventLoopRunner_Action) handleNewGitOpsDeplEvent(ctx context.
 	}
 
 	waitForOperation := !a.testOnlySkipCreateOperation // if it's for a unit test, we don't wait for the operation
-	k8sOperation, dbOperation, err := operations.CreateOperation(ctx, waitForOperation, dbOperationInput,
+	k8sOperation, dbOperation, err := operations.CreateOperation(context.Background(), waitForOperation, dbOperationInput,
```
[the remark this answers is not in the document]
Comment (chetan-rns) [reply backend/eventloop/application_event_loop/application_event_runner_deployments.go]: I've removed the cluster name key from the existing `ctx` before using it for operations. With this approach, we don't have to create a new background context

**Pull request merged**